python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -84,48 +113,59 @@ def __init__(
         self.serialization_ctx = ctx.serialization_context
         assert self.serialization_ctx is not None
 
-    @classmethod
-    def from_channels(
-        cls,
-        gpu_data_channel: "TorchTensorNcclChannel",
-        cpu_data_channel: Optional["Channel"],
-    ):
-        return cls(
-            writer=None,
-            readers=None,
-            gpu_data_typ=None,
-            cpu_data_typ=None,
-            _gpu_data_channel=gpu_data_channel,
-            _cpu_data_channel=cpu_data_channel,
-        )
-
     def __reduce__(self):
         return (
-            NestedTorchTensorNcclChannel.from_channels,
-            (self._gpu_data_channel, self._cpu_data_channel),
+            TorchTensorNcclChannel,
+            (
+                self._writer,
+                self._readers,
+                self._tensor_data_channel,
+                self._non_tensor_data_channel,
+                self._static_non_tensor_data,
+            ),
         )
 
     def ensure_registered_as_writer(self):
```

Comment:
Do we need to call this function in `def write`? Currently, we rely on the underlying channels (i.e., `_tensor_data_channel` and `_non_tensor_data_channel`'s `ensure_xxx` functions) being idempotent. Do we need to add `self._writer_registered` to ensure this function is always idempotent?

```python
def ensure_registered_as_writer(self):
        if self._writer_registered:
            return
        ...
        self._writer_registered = True
```

Reply:
Each channel implementation needs to make sure that their functions are idempotent.